src/core/search/vector.cc

```diff
@@ -5,20 +5,59 @@
 #include "core/search/vector.h"
 
 #include <cmath>
+#include <memory>
 
 #include "base/logging.h"
 
 namespace dfly::search {
 
 using namespace std;
 
+namespace {
+
 // Euclidean vector distance: sqrt( sum: (u[i] - v[i])^2  )
-__attribute__((optimize("fast-math"))) float VectorDistance(const FtVector& u, const FtVector& v) {
-  DCHECK_EQ(u.size(), v.size());
+__attribute__((optimize("fast-math"))) float L2Distance(const float* u, const float* v,
+                                                        size_t dims) {
   float sum = 0;
-  for (size_t i = 0; i < u.size(); i++)
+  for (size_t i = 0; i < dims; i++)
     sum += (u[i] - v[i]) * (u[i] - v[i]);
   return sqrt(sum);
 }
 
+__attribute__((optimize("fast-math"))) float CosineDistance(const float* u, const float* v,
+                                                            size_t dims) {
+  float sum_uv = 0, sum_uu = 0, sum_vv = 0;
+  for (size_t i = 0; i < dims; i++) {
+    sum_uv += u[i] * v[i];
+    sum_uu += u[i] * u[i];
+    sum_vv += v[i] * v[i];
+  }
+
+  if (float denom = sum_uu * sum_vv; denom != 0.0f)
+    return sum_uv / sqrt(denom);
+  return 0.0f;
+}
+
+}  // namespace
+
+OwnedFtVector BytesToFtVector(string_view value) {
+  DCHECK_EQ(value.size() % sizeof(float), 0u) << value.size();
+
+  size_t size = value.size() / sizeof(float);
+  auto out = make_unique<float[]>(size);
```

Comment:
why do you copy? why not just cast?

Reply:
Ah, I forgot to keep the comment when I moved this function.

Because value might be misaligned, i.e. `value.data() % sizeof(float) != 0`. Misaligned pointer access is UB afaik

Reply:
yes, please add the comment and a TODO. I am surprised that this was not shown as a CPU hotspot in your profiling.

1. Vectors are unlikely be stored inside listpacks (too big)
2. And in hashes they are stored as sds values that are indeed misaligned. but we could do magic inside stringmap to allow aligned values.

Reply:
It didn't show up because its called only while indexing, for queries its called only once (for the target vector)

Reply:
Does it mean that if we solve alignment issue, we will reduce the indexing memory by referencing vectors instead of duplicating them?

Reply:
Yes, but it can be fixed in many ways. Its not a problem, I just didn't care yet for optimizing it

For example, we can let `GetVector()` write its output directly to a destination address